man/ash_log.1
.\"
.\"Copyright 2011 Carl Anderson
.\"
.\"Licensed under the Apache License, Version 2.0 (the "License");
.\"you may not use this file except in compliance with the License.
.\"You may obtain a copy of the License at
.\"
.\"    http://www.apache.org/licenses/LICENSE-2.0
.\"
.\"Unless required by applicable law or agreed to in writing, software
.\"distributed under the License is distributed on an "AS IS" BASIS,
.\"WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.\"See the License for the specific language governing permissions and
.\"limitations under the License.
.\"

.TH ash_log 1 \
  "Updated: __DATE__" \
  "__VERSION__" \
  "Advanced Shell History"


.SH NAME
ash_log - The advanced shell history command logger.


.SH SYNOPSIS
Usage: ash_log [options]
      --help
  -a  --alert VALUE
  -c  --command VALUE
  -e  --command_exit VALUE
  -p  --command_pipe_status VALUE
  -s  --command_start VALUE
  -f  --command_finish VALUE
  -n  --command_number VALUE
  -x  --exit VALUE
  -V  --version
  -S  --get_session_id
  -E  --end_session


.SH DESCRIPTION
.B ash_log is not intended to be executed manually.
It is invoked automatically by the advanced shell history system which is 
only activated by sourcing the appropriate file.

Advanced Shell History includes:
.RS
Session details:
.RS
.IP "* login and logout times"
.IP "* username and hostname"
.IP "* incoming ssh connection details (ip and ports)"
.IP "* shell name"
.IP "* tty, shell pid and ppid"
.RE

Command details:
.RS
.IP "* start and stop times"
.IP "* exit code and pipestatus codes"
.IP "* current working directory"
.IP "* session id"
.RE

Many more data points are also collected.
.RE


If you intend to use this system to manage your own personal command history,
it is recommended that you add the appropriate source command to your shell rc
file so it is automatically started with each new session.

.B For bash, add this to your ~/.bashrc file:
.RS
source /usr/lib/advanced_shell_history/bash
.RE

.B For zsh, add this to your ~/.zshrc file:
.RS
source /usr/lib/advanced_shell_history/zsh
.RE


.SH OPTIONS
.IP "      --help"

Display command help and exit 0.

.IP "  -a  --alert VALUE"

Display an alert message (VALUE) and exit 0.

.IP "  -c  --command VALUE"

Store the command (VALUE) entered by the user.

.IP "  -e  --command_exit VALUE"

Store the command exit code (VALUE) of the command.

.IP "  -p  --command_pipe_status VALUE"

Store the command pipe states (exit code for each piped command) separated by
underscores.

.IP "  -s  --command_start VALUE"

The unix epoch timestamp (VALUE) when the command started (the user pressed
Enter).

.IP "  -f  --command_finish VALUE"

The unix epoch timestamp (VALUE) when the command completed and the next prompt
was displayed.

.IP "  -n  --command_number VALUE"

The shell builtin history number (VALUE) of the entered command.

.IP "  -x  --exit VALUE"

The exit code to use when exiting this program.
This flag is needed to re-create the same exit status as the logged command.
This is important because this command runs automatically before the prompt is
redrawn, so if ash_log does not restore the previous exit code, the user would
see the exit code for ash_log (and NOT their command) when executing:
.RS
  echo ${?}
.RE

.IP "  -V  --version"

Display the version number and exit.

.IP "  -S  --get_session_id"

Display the advanced shell history session number and exit.  If no session ID
is found, one is created and displayed.

.IP "  -E  --end_session"

Ends the current session, as defined by the shell environment variable
ASH_SESSION_ID.  It is an error to use this flag without having the
ASH_SESSION_ID variable set.


.SH FILES
.I /etc/ash/ash.conf
.RS
Contains environment variables to be sourced into your shell.
.RE

.I /etc/ash/queries
.RS
Contains saved queries which can be invoked using the
.BR ash_query(1)
command.  Also see
.I ~/.ash/queries
for user-written queries.
.RE

.I ~/.ash/history.db
.RS
The default location for the sqlite3 database holding command history.  See
http://code.google.com/p/advanced-shell-history/w for more details on how
the data is stored internally and other tips for querying the data.
.RE

.I /usr/lib/advanced_shell_history/bash
.RS
Sourced for bash sessions.
.RE

.I /usr/lib/advanced_shell_history/zsh
.RS
Sourced for zsh sessions.
.RE


.SH ENVIRONMENT
.IP ASH_CFG_DB_FAIL_RANDOM_TIMEOUT
After a failed insert, sleep a random number of milliseconds before retrying.
This is intended to add some noise to the retry mechanism.

.IP ASH_CFG_DB_FAIL_TIMEOUT
After a failed insert, sleep this many milliseconds before retrying.

.IP ASH_CFG_DB_MAX_RETRIES
Quit db retries after this many failed attempts.

.IP ASH_CFG_HIDE_USAGE_FOR_NO_ARGS
Normally, if you invoke ash_query with no arguments, the --help output is
displayed.  With this set to a non-empty value, the --help output is
suppressed in this case.

.IP ASH_CFG_HISTORY_DB
The default database to query.  This is set by sourcing one of the shell
scripts in /usr/lib/advanced_shell_history and signifies the location
of the database where commands are logged.  If this variable exists, the
--database flag does not need to be used.

.IP ASH_CFG_IGNORE_UNKNOWN_FLAGS
Normally ash_query complains when it sees unknown flags.  With this variable
set to a non-empty value, unknown flags are ignored.

.IP ASH_CFG_LOG_DATE_FMT
If logging is in use, this format string can be set to customize the date
string.

.IP ASH_CFG_LOG_FILE
The file destination of logged messages, if logging is in use.

.IP ASH_CFG_LOG_IPV4
Can be used to skip logging ipv4 host IP addresses.

.IP ASH_CFG_LOG_IPV6
Can be used to skip logging ipv6 host IP addresses.

.IP ASH_CFG_LOG_LEVEL
The lowest level of logging to make visible.  Levels (in increasing order)
are DEBUG, INFO, WARN, ERROR and FATAL.

.IP ASH_CFG_SKIP_LOOPBACK
Skip logging IP addresses for loopback devices (both ipv4 and ipv6).

.IP ASH_DISABLED
If set, ash_log is disabled.

.IP ASH_SESSION_ID
The session id number created for the current session.  If unset, a new
session ID will be created and displayed.  The caller is expected to export
this variable using the generated ID number.


.SH "SEE ALSO"
.BR ash_query(1)
to query history


.SH AUTHOR
Carl Anderson, Google Inc.


.SH BUGS
Report bugs at http://code.google.com/p/advanced-shell-history/issues


src/ash_log.cpp
/*
   Copyright 2011 Carl Anderson

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

#include "ash_log.hpp"

#include "command.hpp"
#include "config.hpp"
#include "flags.hpp"
#include "logger.hpp"
#include "session.hpp"
#include "unix.hpp"

#include <stdlib.h>  /* for exit, getenv */

#include <iostream>  /* for cerr, cout, endl */
#include <sstream>   /* for stringstream */


DEFINE_string(alert, 'a', 0, "A message to display to the user.");
DEFINE_string(command, 'c', 0, "The command to log.");
DEFINE_int(command_exit, 'e', 0, "The exit code of the command to log.");
DEFINE_string(command_pipe_status, 'p', 0, "The pipe states of the command to log.");
DEFINE_int(command_start, 's', 0, "The timestamp when the command started.");
DEFINE_int(command_finish, 'f', 0, "The timestamp when the command stopped.");
DEFINE_int(command_number, 'n', 0, "The builtin shell history command number.");
DEFINE_int(exit, 'x', 0, "The exit code to use when exiting.");

DEFINE_flag(version, 'V', "Prints the version and exits.");
DEFINE_flag(get_session_id, 'S', "Emits the session ID (or creates one).");
DEFINE_flag(end_session, 'E', "Ends the current session.");


using namespace ash;
using namespace flag;
using namespace std;


/**
 * Displays a brief message about how this is supposed to be used.
 */
void usage(ostream & out) {
  out << "\n\nThis program is not intended to be executed manually.\n\n"
      << "NOTE: See the man page for more details.\n";
  Flag::show_help(out);
  exit(1);
}


int main(int argc, char ** argv) {
  if (getenv("ASH_DISABLED")) return 0;

  // Load the config from the environment.
  Config & config = Config::instance();

  // Log the complete command, if debugging.
  stringstream ss;
  ss << "argv = '[0]='" << argv[0] << "'";
  for (int i = 1; i < argc; ++i)
    ss << ",[" << i << "]='" << argv[i] << "'";
  LOG(DEBUG) << ss.str();

  // Show usage if executed with no args.
  if (argc == 1 && !config.sets("HIDE_USAGE_FOR_NO_ARGS")) {
    Flag::parse(&argc, &argv, true);  // Sets the prog name in help output.
    usage(cerr);
  }

  // Parse the flags.
  Flag::parse(&argc, &argv, true);

  // Display the version and stop: -V
  if (FLAGS_version) {
    cout << ASH_VERSION << endl;
    return 0;
  }

  // Display a user alert: -a 'My alert'
  if (!FLAGS_alert.empty()) {
    cerr << FLAGS_alert << endl;
  }

  // Get the filename backing the history database.
  string db_file = config.get_string("HISTORY_DB");
  if (db_file == "") {
    usage(cerr << "\nExpected ASH_CFG_HISTORY_DB to be defined.");
  }

  // Register the tables expected in the program.
  Session::register_table();
  Command::register_table();

  // Emit the current session number, inserting one if none exists: -S
  if (FLAGS_get_session_id) {
    Database db = Database(db_file);
    stringstream ss;
    char * id = getenv(ASH_SESSION_ID);
    if (id) {
      ss << "select count(*) from sessions where id = " << id
         << " and duration is null;";
      if (db.select_int(ss.str()) == 0) {
        cerr << "ERROR: session_id(" << id << ") not found, "
             << "creating new session." << endl << ss.str() << endl;
        id = 0;
      }
      ss.str("");
    }

    if (id) {
      cout << id << endl;
    } else {
      Session session;
      cout << db.select_int(session.get_sql()) << endl;
    }
  }

  // Insert a command into the DB if there's a command to insert.
  const bool command_flag_used = !FLAGS_command.empty()
    || FLAGS_command_exit
    || !FLAGS_command_pipe_status.empty()
    || FLAGS_command_start
    || FLAGS_command_finish
    || FLAGS_command_number;

  if (command_flag_used) {
    Database db = Database(db_file);
    Command com(FLAGS_command, FLAGS_command_exit, FLAGS_command_start,
      FLAGS_command_finish, FLAGS_command_number, FLAGS_command_pipe_status);
    db.exec(com.get_sql());
  }

  // End the current session in the DB: -E
  if (FLAGS_end_session) {
    char * id = getenv(ASH_SESSION_ID);
    if (id == NULL) {
      LOG(ERROR) << "Can't end the current session: ASH_SESSION_ID undefined.";
    } else {
      Session session;
      Database db = Database(db_file);
      db.exec(session.get_close_session_sql());
    }
  }

  // Set the exit code to match what the previous command exited: -e 123
  return FLAGS_exit;
}

